### DISCUSSION TITLE
125

### DISCUSSION THREAD

**Discussion by giscus** [CODe5753/giscus#5] [Comments]
# 125

서론 실무에서 코드리뷰를 하다가 String 함수 관련해서 개발자 의도와 다르게 작동할 수 있는 코드를 발견했습니다. 이 부분에 대해 바로 피드백드리고 정정할 수 있도록 가이드를 제공해 드렸는데요. 어떤 부분인지 어떻게 올바른 코드를 작성할 수 있는지 알아보겠습니다. Java 17, SpringBoot3를 기준으로 작성되었습니다. 배경 현재 회사에서 수개월의 설계를 끝낸 후 아주 활발하게 개발이 이뤄지고 있습니다. 저와 함께 협력하는 개발자는 같은 회사 동료도 있고 다른 회사에서 협력 중이신 분들도 계십니다. 현재 개발에 들어간 지 겨우 1개월 정도인데요. (3월에 시작) 모든 동료분들이 Java/Spring에 익숙한 상황이 아니기에 각 파트별로 기술적 논의와 표준을 정의하는 담당자를 배정했고 제가 그 ..

https://imksh.com/125

**Comment by haero77** [comment]
얼마 전 회사에서 NPE를 막기 위해 Optional의 정적 팩토리 메서드를 사용하는 것이 좋은가? 에 대한 열띤 논의를 진행했었는데,
이와 관련된 내용이 본문에 포함되어 있어 포스팅을 재밌게 읽었습니다. 

당시 논의에서, 'Optional을 쓰지 않고 직접 null 체크 또는 Objects.isNull을 사용 하는 것이 낫다.'고 결론을 내린 바 있어서, 
관련하여 승현님의 의견도 여쭙고 싶어 댓글 남겨봅니다..! 

---

논의 당시 제가 제시한 의견인 'NPE를 방지하기 위한 null 체크 시 Optional.ofNullable 을 사용 안 하는 것이 좋다.' 의 근거는 다음과 같았습니다.

1. Optioanl의 설계 의도에 반한다

Optioanal의 설계자 Brian Goetz는 스택오버플로우에서 직접 [Optioanl은 메서드의 리턴 타입으로 제한적으로 설계되었음](https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type)을 밝힌 바 있습니다.

다만 그럼 '설계 의도에만 맞게 사용해야 하는가?'를 고민해보자면, 개인적으로는 상황에 맞게 효용성을 고려하는 것을 선호해서 설계 의도에 FIT하진 않을 수 있지만 그것이 개발의 편의성(유지보수성, 가독성 등..)을 가져다준다면 사용해도 괜찮지 않을까라고 생각합니다. 즉, 본문에서 소개된 NPE를 효과적으로 방지하는 방법이 Optional.ofNullable을 사용하는 것이라면 그건 그것대로 좋은 것 아닐까라고 생각했기 때문에, 제가 처음에 제시한 이 근거는 제 스스로를 납득시키기에도 부족해지고 말았습니다....🤣

2. 단순히 값을 얻기 위한 Optional의 정적 팩토리 메서드 사용은 안티 패턴.

`boolean isBlank = Optional.ofNullable(str).map(String::isBlank).orElse(true);`를 천천히 뜯어보면, 
결국 boolean 값을 얻기 위해 팩토리 메서드를 호출하여 Optional 객체를 생성하고, Optional 객체가 제공하는 map과 orElse를 사용하는 형태입니다.

저는 여기서 단순히 값을 얻기 위해서 객체를 생성하는 것이 조금 마음에 걸리더군요. 
물론 요새 컴퓨터야 워낙 성능이 좋으니까 Optioanal 객체 조금 생성한다고 해도 별 영향 없겠지만, 
그래도 객체를 생성하지 않고 간단히 풀어낼 수 있으면 그 편이 더 낫다라고 느꼈기 때문입니다. (본문에서 제시하는 `boolean isBlank = Objects.isNull(str) || str.isBlank();` 같은 방법으로 대체하면 되니까요.) 

관련해서 [DZone 아티클 - 26 Reasons Why Using Optional Correctly Is Not Optional](https://dzone.com/articles/using-optional-correctly-is-not-optional)의 `Item 12: Avoid Chaining Optional's Methods With the Single Purpose of Getting a Value` 항목을 살펴보셔도 좋을 것 같습니다 :)

---

위와 관련해서 승현님은 어떻게 생각하시는지 궁금합니다..! 
긴 글 읽어주셔서 감사합니다!

**Reply by CODe5753** [reply]
@haero77 님 안녕하세요!

Optional 클래스와 관련하여 흥미로운 논의를 하셨었군요. 깊이 있는 내용을 공유해주셔서 감사합니다. 

결국 trade-off 상황인 것 같아요. 
제 결론부터 말씀드리자면 **"Optional 사용이 코드 명료성, 유지보수성 향상을 시키지 않는다면 직관적인 대안을 선택하자"** 입니다.

아시다시피 Optional이 추가적인 래퍼로서의 역할을 하게 되어, 오히려 코드의 복잡성을 증가시킬 수 있는데요. 
실제로 값 접근을 위해 Optional을 사용하는 것은, 말씀하신 것처럼 안티 패턴으로 볼 수 있습니다. 
값 추출을 위해 Optional 객체를 생성하고 여러 단계를 거치는 것은 간단한 null 체크나 다른 조건 검사로 대체될 수 있으며, 이는 성능 측면에서도 더 효율적일 수 있을거에요.

개인적으로는 개발에서 항상 중요한 것은 '**수단(도구)을 적절한 상황에 맞게 사용하는 것**'입니다. 
Optional을 사용할 때도 이 원칙이 적용되어야 한다고 생각합니다. 단순히 NPE를 방지하는 목적으로 Optional을 사용하는 것보다는, 메서드가 반환할 수 있는 값의 부재를 명확하게 표현할 필요가 있을 때 사용하는 것이 적절합니다. 
이렇게 하면, 메서드의 사용자는 명시적으로 Optional의 존재를 확인하고 그에 따라 적절히 대응할 수 있을테니까요.

제가 일하고 있는 환경을 예로 들자면,
코드리뷰가 세세하게 이뤄지고 있지 않아서 널 체크를 빠트릴 확률이 높아 장애로 이어질 수 있는 환경입니다.  
개발자 개개인이 갖고 있는 개발 역량도 천차만별이어서 더욱 어려운 상황이에요.(파트너사와 계약되어 있어서 소속 회사도 다 다르거든요)  
레이어 구조를 사용하고 있기에 Repository 레이어에서 **조회 반환 값을 모두 Optional로 래핑**한다면 이런 장애 상황을 어느정도 방지할 수 있을거에요.  
이걸 컨벤션으로 깔고 간다면 개발자가 할 수 있는 실수를 막아줄 수 있겠지요!

그래서 저는 제가 일하고 있는 환경에서는 Optional 사용을 컨벤션으로 깔아두려고 하고 있어요.

@haero77 님이 일하고 계신 환경에서의 선택과 제가 일하고 있는 곳에서의 선택이 서로 다르지만, 각자 환경에서는 그것이 정답일 수 있다고 생각합니다 :)

충분한 답변이 되었길 바라며,
다시 한번, 공유해주셔서 감사드립니다. 😊

**Comment by Denia-park** [comment]
저도 종종 실수했던 패턴이네요 ㅎㅎ

저도 몇번의 실수 이후로는 `Utils` 함수를 많이 쓰고 있어요.
저는 보통 `org.springframework.util.StringUtils`의 `hasText()` 를 많이 쓰고 있습니다!

**Reply by CODe5753** [reply]
@Denia-park 안녕하세요 답을 이제서야 하네요  
잘 지내고 계시죠? ㅎㅎ 
`hasText`도 유용한 기능이네요! 공유 감사합니다 ㅎㅎ
